File: backend/src/main/java/com/memeit/backend/MyCallBack.java
package com.memeit.backend;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

import static com.memeit.backend.OnCompleteListener.Error.NETWORK_ERROR;
import static com.memeit.backend.OnCompleteListener.Error.OTHER_ERROR;
import static com.memeit.backend.Utils.checkAndFireError;
import static com.memeit.backend.Utils.checkAndFireSuccess;

public  class MyCallBack<T> implements Callback<T> {
    OnCompleteListener<T> listener;

    public MyCallBack(OnCompleteListener<T> listener) {
        this.listener = listener;
    }


    @Override
    public void onResponse(Call<T> call, Response<T> response) {
        if (response.isSuccessful()){
            checkAndFireSuccess(listener,response.body());
        }else{
            checkAndFireError(listener,OTHER_ERROR.setMessage(response.message()));
        }
    }

    @Override
    public void onFailure(Call<T> call, Throwable throwable) {
        Utils.checkAndFireError(listener, NETWORK_ERROR.setMessage(throwable.getMessage()));
    }
}

File: backend/src/main/java/com/memeit/backend/kotlin/MyCallback.kt
package com.memeit.backend.kotlin

import com.memeit.backend.dataclasses.User
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class MyCallback<T>(val onSuccess:(T)->Unit, val onFailure:(String,Call<T>)->Unit) : Callback<T> {
    override fun onResponse(p0: Call<T>?, p1: Response<T>?) {
        onSuccess(p1?.body()!!)
    }

    override fun onFailure(p0: Call<T>?, p1: Throwable?) {
        onFailure(p1?.message!!,p0!!)

    }

}